File: src/test/java/com/scs/CreateNewFileTest.java
package com.scs;

import com.scs.voxlib.VLVoxFile;
import com.scs.voxlib.VLVoxWriter;
import com.scs.voxlib.VLVoxel;
import com.scs.voxlib.chunk.VLVoxGroupChunk;
import com.scs.voxlib.chunk.VLVoxRGBAChunk;
import com.scs.voxlib.chunk.VLVoxRootChunk;
import com.scs.voxlib.chunk.VLVoxShapeChunk;
import com.scs.voxlib.chunk.VLVoxSizeChunk;
import com.scs.voxlib.chunk.VLVoxTransformChunk;
import com.scs.voxlib.chunk.VLVoxXYZIChunk;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;

import static java.nio.file.StandardOpenOption.CREATE;
import static java.nio.file.StandardOpenOption.TRUNCATE_EXISTING;

public class CreateNewFileTest {
    /**
     * This test creates a minimal .vox file.
     */
    @Test
    public void testCreateNewFile() throws IOException {
        // Set size of the model
        var size = new VLVoxSizeChunk(3, 3, 3);

        // Color indices that we'll use. Must be between 1..255.
        final byte ID_GREEN = 1;
        final byte ID_YELLOW = 2;
        final byte ID_RED = 3;

        // Set actual ARGB values for our color indices.
        var paletteArray = new int[4];
        paletteArray[ID_GREEN] = 0xFF33CC33;
        paletteArray[ID_YELLOW] = 0xFFCCCC33;
        paletteArray[ID_RED] = 0xFFCC3333;
        var palette = new VLVoxRGBAChunk(paletteArray);

        // Set voxels using the color indices.
        var voxels = new ArrayList<VLVoxel>();
        voxels.add(new VLVoxel(1, 1, 0, ID_GREEN));
        voxels.add(new VLVoxel(1, 1, 1, ID_YELLOW));
        voxels.add(new VLVoxel(1, 1, 2, ID_RED));
        var model = new VLVoxXYZIChunk(voxels);

        // The following chunks are the necessary containers for our model:
        var groupTransform = new VLVoxTransformChunk(0);
        var group = new VLVoxGroupChunk(1);
        groupTransform.childNodeId = group.id;

        var shapeTransform = new VLVoxTransformChunk(2);
        var shape = new VLVoxShapeChunk(3);
        shape.model_ids.add(0); // id of the 1st model
        shapeTransform.childNodeId = shape.id;
        group.childIds.add(shapeTransform.id);

        // Assemble all our chunks under the root chunk.
        // The order of chunks is important.
        var root = new VLVoxRootChunk();
        root.appendChunk(size);
        root.appendChunk(model);
        root.appendChunk(groupTransform);
        root.appendChunk(group);
        root.appendChunk(shapeTransform);
        root.appendChunk(shape);
        root.appendChunk(palette);
        var voxFile = new VLVoxFile(VLVoxWriter.VERSION, root);

        // Write out the file.
        var path = Paths.get("target/test_file.vox");
        try (
            var outputStream = Files.newOutputStream(path, CREATE, TRUNCATE_EXISTING);
            var writer = new VLVoxWriter(outputStream)
        ) {
            writer.write(voxFile);
        }
        // Magica Voxel should be able to open 'test_file.vox'
    }
}

File: src/test/kotlin/fr/chsfleury/kvox/impl/CreateNewFileTest.kt
package fr.chsfleury.kvox.impl

import fr.chsfleury.kvox.ColorARGB
import fr.chsfleury.kvox.Vec3
import fr.chsfleury.kvox.Voxel
import fr.chsfleury.kvox.chunk.VoxGroupChunk
import fr.chsfleury.kvox.chunk.VoxRGBAChunk
import fr.chsfleury.kvox.chunk.VoxRootChunk
import fr.chsfleury.kvox.chunk.VoxShapeChunk
import fr.chsfleury.kvox.chunk.VoxSizeChunk
import fr.chsfleury.kvox.chunk.VoxTransformChunk
import fr.chsfleury.kvox.chunk.VoxXYZIChunk
import org.junit.jupiter.api.Test
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.nio.file.StandardOpenOption

class CreateNewFileTest {

    /**
     * This test creates a minimal .vox file.
     */
    @Test
    @Throws(IOException::class)
    fun testCreateNewFile() {

        // Set size of the model
        val sizeChunk = VoxSizeChunk(3, 3, 8)

        val green = ColorARGB.of(-0xcc33cd)
        val yellow = ColorARGB.of(-0x3333cd)
        val red = ColorARGB.of(-0x33cccd)
        val blue = ColorARGB(0, 0, 255, 64)
        val colorPalette = MutableColorPalette(
            green,
            yellow,
            red,
            blue
        )

        val palette = VoxRGBAChunk(colorPalette)


        // Set voxels using the color indices.
        val voxels = mutableListOf<Voxel>()
        voxels.add(Voxel(1, 1, 0, colorPalette[green]))
        voxels.add(Voxel(1, 1, 1, colorPalette[yellow]))
        voxels.add(Voxel(1, 1, 2, colorPalette[red]))
        voxels.add(Voxel(1, 1, 3, colorPalette[blue]))
        voxels.add(Voxel(1, 1, 4, colorPalette[red]))
        voxels.add(Voxel(1, 1, 5, colorPalette[yellow]))
        voxels.add(Voxel(1, 1, 6, colorPalette[red]))
        voxels.add(Voxel(1, 1, 7, colorPalette[blue]))
        val model = VoxXYZIChunk(voxels)

        // The following chunks are the necessary containers for our model:
        val shape = VoxShapeChunk(3, listOf(0)) // id of the 1st model
        val shapeTransform = VoxTransformChunk(2, shape.id, Vec3(0, 0, sizeChunk.size.z / 2))

        val shape2 = VoxShapeChunk(5, listOf(0)) // id of the 1st model
        val shapeTransform2 = VoxTransformChunk(4, shape2.id, Vec3(2, 0, sizeChunk.size.z / 2))

        val group = VoxGroupChunk(1, listOf(shapeTransform.id, shapeTransform2.id))
        val groupTransform = VoxTransformChunk(0, group.id)

        // Assemble all our chunks under the root chunk.
        // The order of chunks is important.
        val root = VoxRootChunk(
            sizeChunk,
            model,
            groupTransform,
            group,
            shapeTransform,
            shape,
            shapeTransform2,
            shape2,
            palette
        )

        val voxFile = DefaultVoxFile(DefaultVoxReader.VOX_FORMAT_VERSION, root)

        // Write out the file.
        val path = Paths.get("target/test_file.vox")
        Files.newOutputStream(path, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING).use { outputStream ->
            DefaultVoxWriter(outputStream).use { writer ->
                writer.write(
                    voxFile
                )
            }
        }
        // Magica Voxel should be able to open 'test_file.vox'
    }
}